### PR TITLE
[Bug Fix]Disable DSD for saving ckpt 

### PR DESCRIPTION
**Pull request by acisseJZhong**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/training/_distributed.py [hunk 1]
```diff
@@ -334,42 +335,62 @@ def gather_cpu_state_dict(
...
+            # upcasting NF4 to original dtype
+            param = param.to(param.dtype)
         if is_rank_zero:
-            return cpu_state_dict
-        else:
-            return {}
+            cpu_state_dict[param_name] = param.cpu()
+        torch.distributed.barrier()
+    return cpu_state_dict
+    # TODO: Fix issues with DSD before uncommenting the following code. See #2313 and #2277.
```
Comment (felipemello1) [inline torchtune/training/_distributed.py:355]: this is ugly, but i guess thats what Evan requested. Can you confirm @ebsmothers ?

**Review by felipemello1 — COMMENTED** [review]
@acisseJZhong , if its not too painful, can you confirm that it also works for meta ckpt?

File: torchtune/training/_distributed.py (hunk 1 above)
Comment (ebsmothers) [reply torchtune/training/_distributed.py:355]: Oh sorry I don't want to have the commented-out code. Can we just go back to the version of these functions prior to #2138? I think the TODO is fine though

**Comment by acisseJZhong** [comment]
Verified it works with Meta checkpoint!

**Review by felipemello1 — APPROVED** [review]
